### DISCUSSION TITLE
Unable to migrate from express-graphql: loads forever

### DISCUSSION THREAD

**Discussion by Zhepu-zingbox** [graphql/graphql-http#84] [Q&A] [answered]
**Screenshot**
<img width="970" alt="image" src="https://user-images.githubusercontent.com/52679694/234164237-300d5d95-bc8b-47b7-9f39-039e15afad89.png">
<img width="975" alt="image" src="https://user-images.githubusercontent.com/52679694/234164298-714e7225-d9ae-4f94-8c13-ef17ec18d866.png">

**Expected Behaviour**
Should return
```
{
    "data": {
        "hello": "world"
    }
}
```

**Actual Behaviour**
Check screenshot, loads forever.

**Debug Information**
I used 2 code snippets: one with `express-graphql`, the other `graphql-http`
```
const express = require("express")
const {createHandler} = require('graphql-http');
const { GraphQLSchema, GraphQLObjectType, GraphQLString } = require("graphql");

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      hello: {
        type: GraphQLString,
        resolve: () => 'world',
      },
    },
  }),
});

const app = express()
  app.use(
  "/graphql",
  createHandler({
    schema
  })
)
app.listen(4000)
console.log("Running a GraphQL API server at http://localhost:4000/graphql")
```

```
const express = require("express")
const { graphqlHTTP } = require("express-graphql")
const { GraphQLSchema, GraphQLObjectType, GraphQLString } = require("graphql");
const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      hello: {
        type: GraphQLString,
        resolve: () => 'world',
      },
    },
  }),
});

const app = express()
app.use(
  "/graphql",
  graphqlHTTP({
    schema: schema
  })
)
app.listen(4000)
console.log("Running a GraphQL API server at http://localhost:4000/graphql")
```

**Further Information**
I'm with `node v16.16.0`
package.json
```
{
  "name": "demo",
  "version": "1.0.0",
  "description": "demo",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "license": "ISC",
  "dependencies": {
    "express": "^4.18.2",
    "express-graphql": "^0.12.0",
    "graphql-http": "^1.18.0"
  }
}
```

**Comment by enisdenjo** [comment, marked as answer]
Hey there, thanks for writing. The issue is that you're not using the express specific handler from graphql-http. Consider making the following change:

```diff
- const {createHandler} = require('graphql-http');
+ const {createHandler} = require('graphql-http/lib/use/express');
```

**Reply by Zhepu-zingbox** [reply]
That's really helpful! Thx a lot!

**Comment by tomwyr** [comment]
@enisdenjo  Hey, just wanted to add my experience with adding `express-graphql` to my project. I spent about an hour before I stumbled upon this topic and realized there's a specific handler per each library that the project supports.
I find this way of importing the handler rather unintuitive and potentially dangerous. The common practice I've seen most of the times is that packages export publicly visible apis in the root export, and importing any internal sources is a potential breaking change at any moment.

That said, I should have really head to the documentation first where I'd find that bit of information.

**Reply by benjie** [reply]
This pattern is becoming more popular as the Node.js ecosystem diversifies, it allows importing something that supports your given library without consuming memory by loading the code (and types!) for all supported libraries. (It's also a common way of working around restrictions in tree shaking by only importing a subset of the module.)

Node.js has added explicit support for controlling the public interface of your module including subpaths via the `"exports"` property in `package.json`; you can read more about it here:

https://nodejs.org/docs/latest/api/packages.html#exports

**Reply by tomwyr** [reply]
I wasn't fully aware of those things about the Node ecosystem. Thanks for filling me in.
